kgx/sink/tsv_sink.py

```diff
@@ -0,0 +1,194 @@
+import os
+import tarfile
+from typing import Optional, Dict, Set, Any
+from ordered_set import OrderedSet
+
+from kgx.sink.sink import Sink
+from kgx.utils.kgx_utils import extension_types, archive_write_mode, archive_format, remove_null, _sanitize_export
+
+
+class TsvSink(Sink):
+    """
+    TsvSink is responsible for writing data as records to a TSV/CSV.
+
+    Parameters
+    ----------
+    filename: str
+        The filename to write to
+    format: str
+        The file format (``tsv``, ``csv``)
+    compression: str
+        The compression type (``tar``, ``tar.gz``)
+    kwargs: Any
+        Any additional arguments
+    """
+
+    def __init__(self, filename: str, format: str, compression: Optional[str] = None, **kwargs: Any):
+        super().__init__()
+        if format not in extension_types:
+            raise Exception(f'Unsupported format: {format}')
+        self.delimiter = extension_types[format]
+        self.dirname = os.path.abspath(os.path.dirname(filename))
+        self.basename = os.path.basename(filename)
+        self.extension = format.split(':')[0]
+        self.mode = archive_write_mode[compression] if compression in archive_write_mode else None
+        self.nodes_file_basename = f"{self.basename}_nodes.{self.extension}"
+        self.edges_file_basename = f"{self.basename}_edges.{self.extension}"
+        if self.dirname:
+            os.makedirs(self.dirname, exist_ok=True)
+
+        self._node_properties = kwargs['node_properties'] if 'node_properties' in kwargs else set()
+        self.ordered_node_columns = TsvSink._order_node_columns(self._node_properties)
+        self._edge_properties = kwargs['edge_properties'] if 'edge_properties' in kwargs else set()
+        self.ordered_edge_columns = TsvSink._order_edge_columns(self._edge_properties)
+
+        self.nodes_file_name = os.path.join(self.dirname if self.dirname else '', self.nodes_file_basename)
+        self.NFH = open(self.nodes_file_name, 'w')
+        self.NFH.write(self.delimiter.join(self.ordered_node_columns) + '\n')
+        self.edges_file_name = os.path.join(self.dirname if self.dirname else '', self.edges_file_basename)
+        self.EFH = open(self.edges_file_name, 'w')
+        self.EFH.write(self.delimiter.join(self.ordered_edge_columns) + '\n')
+
+    def write_node(self, record: Dict) -> None:
+        """
+        Write a node record to the underlying store.
+
+        Parameters
+        ----------
+        record: Dict
+            A node record
+
+        """
+        row = self._build_export_row(record)
+        row['id'] = record['id']
+        values = []
+        for c in self.ordered_node_columns:
+            if c in row:
+                values.append(str(row[c]))
```

Comment:
unless I'm mistaken, this doesn't escape tabs/newlines, which could cause havoc in TSVs

Reply:
Good catch!

The tabs/newlines are removed at the time of reading records. i.e. By the time we get here, the errant tabs/newlines are removed. But of course, this has to be tested properly.